nifi-nar-bundles/nifi-sql-reporting-bundle/nifi-sql-reporting-tasks/src/main/java/org/apache/nifi/reporting/sql/MetricsSqlQueryService.java

```diff
@@ -0,0 +1,220 @@
+                        throw new IOException("Failed to close statement", e);
+                    }
+                }
+            }
+
+            @Override
+            public ResultSet getResultSet() {
+                return rs;
+            }
+
+            @Override
+            public int getRecordsRead() {
+                return 0;
+            }
+
+        };
+    }
+
+    public ResultSetRecordSet getResultSetRecordSet(QueryResult queryResult) throws Exception{
+
+        final ResultSet rs = queryResult.getResultSet();
+        ResultSetRecordSet recordSet = null;
+        // Create the RecordSchema from the ResultSet (same way ExecuteSQL does it)
+        final RecordSchema writerSchema = AvroTypeUtil.createSchema(JdbcCommon.createSchema(rs));
+
+        try {
+            recordSet = new ResultSetRecordSet(rs, writerSchema);
+        } catch (final SQLException e) {
+            getLogger().error("Error creating record set from query results due to {}", new Object[]{e.getMessage()}, e);
+        }
+
+        return recordSet;
+    }
+
+    private synchronized CachedStatement getStatement(final String sql, final Supplier<CachedStatement> statementBuilder, Cache<String, BlockingQueue<CachedStatement>> statementQueues) {
+        final BlockingQueue<CachedStatement> statementQueue = statementQueues.get(sql, key -> new LinkedBlockingQueue<>());
+
+        if(statementQueue != null) {
+            final CachedStatement cachedStmt = statementQueue.poll();
+            if (cachedStmt != null) {
+                return cachedStmt;
+            }
+        }
+
+        return statementBuilder.get();
+    }
+
+    private CachedStatement buildCachedStatement(final String sql, final ReportingContext context) {
+
+        final CalciteConnection connection = createConnection();
+        final SchemaPlus rootSchema = createRootSchema(connection);
+
+        final ConnectionStatusTable connectionStatusTable = new ConnectionStatusTable(context, getLogger());
+        rootSchema.add("CONNECTION_STATUS", connectionStatusTable);
+        final ConnectionStatusPredictionsTable connectionStatusPredictionsTable = new ConnectionStatusPredictionsTable(context, getLogger());
+        rootSchema.add("CONNECTION_STATUS_PREDICTIONS", connectionStatusPredictionsTable);
+        final ProcessorStatusTable processorStatusTable = new ProcessorStatusTable(context, getLogger());
+        rootSchema.add("PROCESSOR_STATUS", processorStatusTable);
+        final ProcessGroupStatusTable processGroupStatusTable = new ProcessGroupStatusTable(context, getLogger());
+        rootSchema.add("PROCESS_GROUP_STATUS", processGroupStatusTable);
+        final JvmMetricsTable jvmMetricsTable = new JvmMetricsTable(context, getLogger());
+        rootSchema.add("JVM_METRICS", jvmMetricsTable);
+        final BulletinTable bulletinTable = new BulletinTable(context, getLogger());
+        rootSchema.add("BULLETINS", bulletinTable);
+
+        // TODO add the others
```

Comment:
@mattyb149 is this TODO still needed?

Reply:
Nope needed to be done, should be up to date now